File: python/sort-the-matrix-diagonally.py
class Solution:
    def get_diagonal(mat: List[List[int]], x: int, y: int) -> List[int]:
        diagonal = []
        while x < len(mat) and y < len(mat[x]):
            diagonal.append(mat[x][y])
            x += 1
            y += 1
        return diagonal

    def fill_diagonal(mat: List[List[int]], x: int, y: int, diagonal: List[int]):
        for i in range(len(diagonal)):
            mat[x + i][y + i] = diagonal[i]

    def sort_diagonal(mat: List[List[int]], x: int, y: int):
        diagonal = Solution.get_diagonal(mat, x, y)
        diagonal.sort()
        Solution.fill_diagonal(mat, x, y, diagonal)

    def diagonalSort(self, mat: List[List[int]]) -> List[List[int]]:
        for x in range(len(mat)):
            Solution.sort_diagonal(mat, x, 0)
        for y in range(1, len(mat[0])):
            Solution.sort_diagonal(mat, 0, y)
        return mat

File: go/sort-the-matrix-diagonally.go
import (
    "sort"
)

func get_diagonal(mat [][]int, x int, y int) []int {
    diagonal := make([]int, 0)
    for x < len(mat) && y < len(mat[x]) {
        diagonal = append(diagonal, mat[x][y])
        x += 1
        y += 1
    }
    return diagonal
}

func fill_diagonal(mat [][]int, x int, y int, diagonal []int) {
    for i := 0; i < len(diagonal); i++ {
        mat[x + i][y + i] = diagonal[i]
    }
}

func sort_diagonal(mat [][]int, x int, y int) {
    diagonal := get_diagonal(mat, x, y)
    sort.Ints(diagonal)
    fill_diagonal(mat, x, y, diagonal)
}

func diagonalSort(mat [][]int) [][]int {
    for x := 0; x < len(mat); x++ {
        sort_diagonal(mat, x, 0)
    }
    for y := 1; y < len(mat[0]); y++ {
        sort_diagonal(mat, 0, y)
    }
    return mat
}
